File: bot.js
const TelegramBot = require('node-telegram-bot-api');
const { Connection, PublicKey } = require('@solana/web3.js');
const { Metadata } = require('@metaplex-foundation/mpl-token-metadata');

// Telegram Bot Token (Replace with your actual bot token)
const BOT_TOKEN = "7420937580:AAFd9Y1fkZZwt7AwnngaLAvl3X91r_G8SNg";

// Solana RPC Endpoint (Use mainnet or devnet as needed)
const SOLANA_RPC = "https://api.mainnet-beta.solana.com";
const connection = new Connection(SOLANA_RPC);

// Initialize Telegram Bot
const bot = new TelegramBot(BOT_TOKEN, { polling: true });

console.log("🤖 Solana Anti-Rug Bot is running...");

// Function to fetch token metadata from the blockchain
async function getTokenMetadata(mintAddress) {
    try {
        const mintPublicKey = new PublicKey(mintAddress);
        const metadataPDA = await Metadata.getPDA(mintPublicKey);
        const metadataAccount = await Metadata.load(connection, metadataPDA);
        const { data } = metadataAccount.data;

        return {
            name: data.name.trim(),
            symbol: data.symbol.trim(),
            uri: data.uri,
            sellerFeeBasisPoints: data.sellerFeeBasisPoints,
        };
    } catch (error) {
        console.error(`Error fetching metadata: ${error.message}`);
        return null;
    }
}

// Telegram bot command to check a token
bot.onText(/\/check (.+)/, async (msg, match) => {
    const chatId = msg.chat.id;
    const tokenMint = match[1].trim();

    try {
        // Validate if the provided mint address is a valid Solana address
        new PublicKey(tokenMint);
    } catch (error) {
        bot.sendMessage(chatId, '❌ Invalid token mint address.');
        return;
    }

    bot.sendMessage(chatId, `🔍 Checking token: ${tokenMint}...`);

    const metadata = await getTokenMetadata(tokenMint);
    if (!metadata) {
        bot.sendMessage(chatId, '❌ Token metadata not found or inaccessible.');
        return;
    }

    let message = `✅ **Token Analysis**\n`;
    message += `🔹 **Name**: ${metadata.name}\n`;
    message += `🔹 **Symbol**: ${metadata.symbol}\n`;
    message += `🔹 **URI**: ${metadata.uri}\n`;
    message += `🔹 **Seller Fee Basis Points**: ${metadata.sellerFeeBasisPoints / 100}%\n`;

    bot.sendMessage(chatId, message, { parse_mode: "Markdown" });
});

File: bot.py
import telebot
from solana.rpc.api import Client
from solana.publickey import PublicKey

# Telegram Bot Token (Replace with your actual bot token)
BOT_TOKEN = "7420937580:AAFd9Y1fkZZwt7AwnngaLAvl3X91r_G8SNg"

# Solana RPC Endpoint
SOLANA_RPC_URL = "https://api.mainnet-beta.solana.com"
client = Client(SOLANA_RPC_URL)

# Initialize Telegram Bot
bot = telebot.TeleBot(BOT_TOKEN)

print("🤖 Solana Anti-Rug Bot is running...")


# Function to fetch token metadata
def get_token_metadata(mint_address):
    try:
        mint_pubkey = PublicKey(mint_address)

        # Fetch token supply info
        supply_info = client.get_token_supply(mint_pubkey)
        if not supply_info["result"]:
            return None

        # Fetch account info
        account_info = client.get_account_info(mint_pubkey)
        if not account_info["result"]["value"]:
            return None

        return {
            "supply": supply_info["result"]["value"]["amount"],
            "decimals": supply_info["result"]["value"]["decimals"],
        }
    except Exception as e:
        print(f"Error fetching metadata: {e}")
        return None


# Telegram bot command to check a token
@bot.message_handler(commands=['check'])
def check_token(message):
    chat_id = message.chat.id
    try:
        token_mint = message.text.split(" ")[1].strip()
        token_metadata = get_token_metadata(token_mint)

        if not token_metadata:
            bot.send_message(chat_id, "❌ Token metadata not found or inaccessible.")
            return

        supply = int(token_metadata["supply"]) / (10 ** token_metadata["decimals"])
        msg = f"✅ **Token Analysis**\n"
        msg += f"🔹 **Supply**: {supply}\n"
        bot.send_message(chat_id, msg, parse_mode="Markdown")

    except IndexError:
        bot.send_message(chat_id, "❌ Please provide a valid token mint address.")
    except Exception as e:
        bot.send_message(chat_id, f"❌ Error: {e}")


# Start polling
bot.polling()
